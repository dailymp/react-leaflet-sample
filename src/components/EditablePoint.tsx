import * as React from "react";
import { Point } from "../common/Point";

interface Props {
    pins: Point[];
}

const marginPointComponentsCSS = () => ({marginLeft: '1em'});

export const EditablePoints = (props: Props) => {
    // TODO: Creative idea: Use hooks to edit data in inputs and when you are finished push in apply changes to map.
    return (
        <>
            <h1>Edit points</h1>
            <button>Apply Changes</button>
            {
                props.pins.map((pin: Point) => (
                    <p>
                        <label >X: </label>
                        <input type="text" value={pin.x} style={marginPointComponentsCSS()} />
                        <label style={marginPointComponentsCSS()}>Y: </label>
                        <input type="text" value={pin.y} style={marginPointComponentsCSS()} />
                        <label style={marginPointComponentsCSS()}>Color: </label>
                        <input type="text" value={pin.hexadecimalColor} style={marginPointComponentsCSS()} />
                        <button style={marginPointComponentsCSS()}>Delete</button>
                    </p>
                ))
            }
        </>
    );
}