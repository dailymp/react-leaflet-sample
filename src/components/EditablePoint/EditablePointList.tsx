import * as React from "react";
import { Point } from "../../common/Point";
import { EditablePoint } from "./EditablePoint";

interface PropsEditablePointList {
    pins: Point[];
    onEditPins: (newPins: Point[]) => void;
    setShowPointList: React.Dispatch<React.SetStateAction<boolean>>;
}

export interface PointWithEnable extends Point {
    isEnable: boolean;
}

export const marginComponentsInEditablePointList = () => ({marginLeft: '1em'});

const parseGetAndSetPinListToPins = (getAndSetPin: [PointWithEnable, React.Dispatch<React.SetStateAction<PointWithEnable>>][]): Point[] => (
    getAndSetPin.map(([pin, _]) => ({
        x: pin.x,
        y: pin.y,
        hexadecimalColor: pin.hexadecimalColor,
    }))
);

export const EditablePointList = (props: PropsEditablePointList) => {
    const getAndSetPin = props.pins.map((pin: Point) => {
        return React.useState({...pin, isEnable: true});
    });
    let indexPin = 0;

    return (
        <>
            <p>
                <button
                    onClick={evt => {
                        props.onEditPins(parseGetAndSetPinListToPins(getAndSetPin));
                        props.setShowPointList(false);
                    }}
                >Apply Changes</button>
                <button
                    style={marginComponentsInEditablePointList()}
                    onClick={evt => props.setShowPointList(false)}
                >Cancel Changes</button>
            </p>
            {
                getAndSetPin.map(([pin, setPin]) => {
                    const result = pin.isEnable ?
                        <EditablePoint
                            key={`editable-point-${indexPin}`}
                            pin={pin}
                            setPin={setPin} />
                        : undefined;
                    indexPin++;
                    return result;
                })
            }
        </>
    );
}

