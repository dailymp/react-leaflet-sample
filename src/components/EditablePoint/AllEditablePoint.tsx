import * as React from "react";
import { Point } from "../../common/Point";
import { EditablePointList } from "./EditablePointList";

interface Props {
    pins: Point[];
    onEditPins: (newPins: Point[]) => void;
}

export const AllEditablePoints = (props: Props) => {
    const [showPointList, setShowPointList] = React.useState(false);

    return(
        <>
        <h1>Edit points</h1>
        {
            showPointList ?
                <>
                    <EditablePointList {...props} setShowPointList={setShowPointList} />
                </>
                : <button onClick={evt => setShowPointList(true)}>Edit Point List</button>
        }
        </>
    );
}
