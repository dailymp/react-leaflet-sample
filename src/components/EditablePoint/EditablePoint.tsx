import * as React from "react";
import { marginComponentsInEditablePointList, PointWithEnable } from "./EditablePointList";

interface PropsPoint {
    getAndSetPin: [PointWithEnable, React.Dispatch<React.SetStateAction<PointWithEnable>>][],
    pin: PointWithEnable,
    setPin: React.Dispatch<React.SetStateAction<PointWithEnable>>,
    indexPin: number
}

export const EditablePoint = (props: PropsPoint) => (
    <p>
        <label >X: </label>
        <input
            type="text"
            value={props.pin.x}
            style={marginComponentsInEditablePointList()}
            onChange={event => props.setPin({...props.pin, x: +event.target.value})}
        />
        <label style={marginComponentsInEditablePointList()}>Y: </label>
        <input
            type="text"
            value={props.pin.y}
            style={marginComponentsInEditablePointList()}
            onChange={event => props.setPin({...props.pin, y: +event.target.value})}
        />
        <label style={marginComponentsInEditablePointList()}>Color: </label>
        <input
            type="text"
            value={props.pin.hexadecimalColor}
            style={marginComponentsInEditablePointList()}
            onChange={event => props.setPin({...props.pin, hexadecimalColor: event.target.value})}
        />
        <button
            style={marginComponentsInEditablePointList()}
            onClick={event => {
                event.currentTarget.disabled = true; // If there's a problem, at least the button is disabled.
                props.setPin({...props.pin, isEnable: false});
            }}
        >Delete</button>
    </p>
);
