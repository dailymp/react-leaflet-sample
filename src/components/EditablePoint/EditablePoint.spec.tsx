import * as React from 'react';
import { shallow } from 'enzyme';
import { EditablePoint } from './EditablePoint';

describe('Tests for EditablePoint component', () => {
  it('should render as expected when passing required properties', () => {
    const props = {
        pin: {x: 1, y:3, hexadecimalColor: '#00e700', isEnable: true},
        setPin: jest.fn()
    };

    const component = shallow(<EditablePoint pin={props.pin} setPin={props.setPin} />);

    expect(component).toMatchSnapshot();
  });
});