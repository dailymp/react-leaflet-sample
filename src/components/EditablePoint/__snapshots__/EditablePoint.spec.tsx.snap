// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for EditablePoint component should render as expected when passing required properties 1`] = `
<p>
  <label>
    X: 
  </label>
  <input
    onChange={[Function]}
    style={
      Object {
        "marginLeft": "1em",
      }
    }
    type="text"
    value={1}
  />
  <label
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    Y: 
  </label>
  <input
    onChange={[Function]}
    style={
      Object {
        "marginLeft": "1em",
      }
    }
    type="text"
    value={3}
  />
  <label
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    Color: 
  </label>
  <input
    onChange={[Function]}
    style={
      Object {
        "marginLeft": "1em",
      }
    }
    type="text"
    value="#00e700"
  />
  <button
    onClick={[Function]}
    style={
      Object {
        "marginLeft": "1em",
      }
    }
  >
    Delete
  </button>
</p>
`;
